### DISCUSSION TITLE
Build Keystone schema from existing Prisma schema

### DISCUSSION THREAD

**Discussion by SheaBelsky** [keystonejs/keystone#7233] [Questions] [answered]
I'm looking to possibly adapt an existing project built with Prisma and TypeGraphQL to fully use Keystone 6. I understand that Keystone 6 is built on top of Prisma, and building Keystone's schema outputs a Prisma schema. However, is it possible to go in the reverse direction if I want to introduce Keystone into a project which already has a Prisma schema? Or do I need to manually construct the Keystone schema and keep building on top of it until the output schema exactly matches my current schema?

Prisma has a feature similar to this, where it can introspect a database and construct a Prisma schema from just a database. I'm not sure if comparable functionality exists for Keystone, but it would certainly make the adoption process for established apps much easier.

Thanks!

**Comment by SheaBelsky** [comment]
Might actually be a duplicate of this (I tried searching before posting but didn't see this one until I searched again: https://github.com/keystonejs/keystone/discussions/6299)

**Comment by dcousens** [comment, marked as answer]
Hi @SheaBelsky,  Keystone 6 is highly opinionated about the underlying database names and structure by default.  However,  there are features like [`@@map`](https://keystonejs.com/docs/apis/schema#db) (see [Prisma documentation](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#map-1)) which might be possible to help you leverage an existing database.

I think the best approach here would be to let Keystone build a database,  and then use a diff between that and your existing database to resolve each mapping until you reach parity (assuming that is actually possible).

I doubt you'll be lucky enough to have an outcome where no migration is required.

**Reply by SheaBelsky** [reply]
Thanks for clarifying this! My biggest blocker here is converting array fields (like an array of enums or array of strings) to JSON fields. I'll keep working to make the migrations and schema changes as minimal as possible, though I doubt it will be perfect. Do you have any insight on the array front?

**Reply by dcousens** [reply]
@SheaBelsky can you post an example of your array fields as they would be configured in Prisma?

**Reply by SheaBelsky** [reply]
Here's an example of an enum I might have in Prisma:

```
jobInPersonLocation        OfficeEnvironment[]

enum OfficeEnvironment {
  INDOORS_SEPARATED
  INDOORS_TOGETHER
  IN_PERSON
  OUTDOORS
  REMOTE
}
```

In this particular situation, `jobInPersonLocation` could be referred to as a dropdown of items. But I would like to also be able to refer to these exact values in other Keystones models in order to avoid copy/pasting them as two separate arrays. Does Keystone take care of comparing string literals such as this and establishing one consistent enum for them? Or would they be treated as two different things?

**Comment by devmanbr** [comment]
@SheaBelsky, did you find any solution for this?

prisma offers the advantage of validating the entire schema at development time, something that is not possible from the keystone schema.

**Comment by SheaBelsky** [comment]
@devmanbr I ended up not pursuing this route as intensely as I had planned - The reason I had gone down this rabbit hole was because I lacked the time to build out a proper API and admin interface for my Prisma schema, and Keystone took care of that for me. I now work on this project full time and use TypeGraphQL to consume the Prisma schema and emit a 1:1 GraphQL API.

https://github.com/MichalLytek/typegraphql-prisma

**Reply by devmanbr** [reply]
the project looks really interesting, @SheaBelsky! it's certainly something to consider for backend.

however, are there any open-source CMS solutions that you know of that don't require the additional effort that keystonejs currently requires?

**Reply by SheaBelsky** [reply]
Prisma has a built-in offering called Prisma studio which accomplishes some of what Keystone accomplishes in terms of the CMS/admin interface.

https://www.prisma.io/studio

Admin.js also offers a similar plugin experience which creates a robust CMS for the Prisma schema:

https://docs.adminjs.co/installation/adapters/prisma
